Improve provide conflict error message

uilianries opened the pull request:
Changelog: Fix: Improve `provides` conflict message error.
Docs: Omit

- Split up old `GraphError` generic class to specialized classes 
- No more consult `.kind` attribute, each class has its own information
- Update tests which depends on graph error
- Improve graph error messages.

fixes #13548

- [ ] Refer to the issue that supports this Pull Request.
- [ ] If the issue has missing info, explain the purpose/use case/pain/need that covers this Pull Request.
- [ ] I've read the [Contributing guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [ ] I've followed the PEP8 style guides for Python code.
- [ ] I've opened another PR in the Conan docs repo to the ``develop`` branch, documenting this one. 

<sup>**Note:** By default this PR will skip the slower tests and will use a limited set of python versions. Check [here](https://github.com/conan-io/conan/blob/develop/.github/PR_INCREASE_TESTING.md) how to increase the testing level by writing some tags in the current PR body text.</sup>

memsharded commented on conans/client/graph/error/conflict_error.py:
```diff
@@ -0,0 +1,15 @@
+from conans.errors import ConanException
+
+
+class GraphConflictError(ConanException):
```
Make all the classes inherit from a common GraphError or something like that.

uilianries replied:
Done.

memsharded commented on conans/client/graph/error/loop_error.py:
```diff
@@ -0,0 +1,15 @@
+from conans.errors import ConanException
```
Better keep all the exceptions in the same file, no need explosion of small files.

uilianries replied:
Done.

memsharded commented on conans/client/graph/graph_builder.py:
```diff
@@ -59,7 +60,8 @@ def load_graph(self, root_node, profile_host, profile_build, graph_lock=None):
             check_graph_provides(dep_graph)
             self._compute_test_package_deps(dep_graph)
-        except GraphError as e:
+        except (GraphRuntimeError, GraphProvidesError, GraphMissingError, GraphRuntimeError,
```
Use general ``GraphError`` exception

uilianries replied:
Done.

memsharded reviewed (commented):
I am good with the refactor of dedicated classes for each error type, please see the minor comments.

uilianries committed:
Refactor GraphError

- Split each class part to a separated class
- Improve output messages for each error
- Update tests to use the new classes
- Update exception catch to use dedicated classes

Signed-off-by: Uilian Ries <uilianries@gmail.com>

memsharded reviewed (approved):
Looks good!

uilianries committed:
Fix super class

Signed-off-by: Uilian Ries <uilianries@gmail.com>

AbrilRBS reviewed (approved):
Nice!
